Structured concurrency

c42f opened the pull request:
A document describing structured concurrency and (eventually) how it might be applied in julia.

**Please direct substantive comments to the JuliaLang/julia issue https://github.com/JuliaLang/julia/issues/33248**

(procedural errors such as spelling may be pointed out here)

c42f committed:
WIP: Structured concurrency

tkf committed:
Add a few more links

c42f committed:
Merge pull request #4 from tkf/links

Add a few more links

c42f committed:
Tweak resource handling

c42f committed:
Fix some capitalization

c42f committed:
Add note on the benefits of async/await syntax

Co-authored-by: Takafumi Arakaki <aka.tkf@gmail.com>

vchuravy commented on StructuredConcurrency.md:
```diff
@@ -0,0 +1,283 @@
+
+## Related julia issues and prototypes
+
+* [Tapir parrallel IR](https://github.com/JuliaLang/julia/pull/31086)
```
Note that this is primarily about:
- Compiler representation of tasks
- Experimenting with Cilk-style semantics (e.g. may happen in parallel or what I have taken to call structured parallelism)

c42f replied:
Thanks, I've replied to this in the main repo.

c42f committed:
Add some metadata

I have used "Editor" here rather than "Author" as this is a very broad
topic and I hope many people will contribute to the discussion.

c42f committed:
Fix a few spelling mistakes

c42f commented:
I have merged this; not because I think it's complete in any way, but to give us a base to work on.

ArchRobison commented on StructuredConcurrency.md:
```diff
@@ -0,0 +1,289 @@
+#### Runtime technicalities for preemption
+
+On a technical level, our runtime situation in julia-1.3 is very similar to
+Go where preemption is cooperative and a rouge goroutine can sometimes wedge
```
Typo: `rouge` -> `rogue`

c42f replied:
Hi Arch, thanks; will be fixed in the next revision. If you do have time to provide some perspective, please head on over to

https://github.com/JuliaLang/julia/issues/33248

where I am pulling on a lot of disparate threads and undoubtedly need all the help I can get from people who actually know things :-)
